### PR TITLE
feat(ESSNTL-4025): Add logs to xjoin-api-gateway

### PR DESCRIPTION
**Pull request by jpramos123**
In this PR we:

1) Add logs to xjoin-api-gateway and increase timeout to fetch data from Schema Registry
2) Create error handlers for the API calls
3) Add lint to the code

Log format (in `debug` mode):

```
{"level":"info","message":"Starting Apollo Server","timestamp":"2023-02-15T19:48:33.940Z"}
{"level":"info","message":"Fetching Subgraph Schemas every 600000ms","timestamp":"2023-02-15T19:48:33.942Z"}
{"level":"info","message":"Fetching artifacts from Schema Registry","timestamp":"2023-02-15T19:48:33.943Z"}
{"level":"debug","message":"Fetching artifacts at: http://apicurio:1080/apis/registry/v2/search/artifacts?labels=graphql","timestamp":"2023-02-15T19:48:33.943Z"}
{"level":"info","message":"Found 4 artifacts","timestamp":"2023-02-15T19:48:33.963Z"}
{"level":"info","message":"Fetching details of artifact: xjoinindexpipeline.hosts-hbi-tags.1675279470601933223","timestamp":"2023-02-15T19:48:33.963Z"}
{"level":"debug","message":"Using resource path: v1/artifacts/xjoinindexpipeline.hosts-hbi-tags.1675279470601933223","timestamp":"2023-02-15T19:48:33.963Z"}
{"level":"debug","message":"Fetching artifact details at: http://apicurio:1080/apis/registry/v1/artifacts/xjoinindexpipeline.hosts-hbi-tags.1675279470601933223","timestamp":"2023-02-15T19:48:33.963Z"}
{"level":"debug","message":"Extracting Subgraph URL from artifact: xjoinindexpipeline.hosts-hbi-tags.1675279470601933223","timestamp":"2023-02-15T19:48:33.970Z"}
{"level":"debug","message":"Extracted URL from artifact: http://xjoinindexpipeline-hosts-hbi-tags-1675279470601933223.test.svc:4000/graphql","timestamp":"2023-02-15T19:48:33.971Z"}
{"level":"info","message":"Fetching details of artifact: xjoinindexpipeline.hosts-hbi-tags.1676329308196796999","timestamp":"2023-02-15T19:48:33.973Z"}
{"level":"debug","message":"Using resource path: v1/artifacts/xjoinindexpipeline.hosts-hbi-tags.1676329308196796999","timestamp":"2023-02-15T19:48:33.974Z"}
{"level":"debug","message":"Fetching artifact details at: http://apicurio:1080/apis/registry/v1/artifacts/xjoinindexpipeline.hosts-hbi-tags.1676329308196796999","timestamp":"2023-02-15T19:48:33.974Z"}
{"level":"debug","message":"Extracting Subgraph URL from artifact: xjoinindexpipeline.hosts-hbi-tags.1676329308196796999","timestamp":"2023-02-15T19:48:33.981Z"}
{"level":"debug","message":"Extracted URL from artifact: http://xjoinindexpipeline-hosts-hbi-tags-1676329308196796999.test.svc:4000/graphql","timestamp":"2023-02-15T19:48:33.982Z"}
{"level":"info","message":"Fetching details of artifact: xjoinindexpipeline.hosts.1675279470601933223","timestamp":"2023-02-15T19:48:33.982Z"}
{"level":"debug","message":"Using resource path: v1/artifacts/xjoinindexpipeline.hosts.1675279470601933223","timestamp":"2023-02-15T19:48:33.982Z"}
{"level":"debug","message":"Fetching artifact details at: http://apicurio:1080/apis/registry/v1/artifacts/xjoinindexpipeline.hosts.1675279470601933223","timestamp":"2023-02-15T19:48:33.982Z"}
{"level":"debug","message":"Extracting Subgraph URL from artifact: xjoinindexpipeline.hosts.1675279470601933223","timestamp":"2023-02-15T19:48:33.989Z"}
{"level":"debug","message":"Extracted URL from artifact: http://xjoinindexpipeline-hosts-1675279470601933223.test.svc:4000/graphql","timestamp":"2023-02-15T19:48:33.989Z"}
{"level":"info","message":"Fetching details of artifact: xjoinindexpipeline.hosts.1676329308196796999","timestamp":"2023-02-15T19:48:33.992Z"}
{"level":"debug","message":"Using resource path: v1/artifacts/xjoinindexpipeline.hosts.1676329308196796999","timestamp":"2023-02-15T19:48:33.992Z"}
{"level":"debug","message":"Fetching artifact details at: http://apicurio:1080/apis/registry/v1/artifacts/xjoinindexpipeline.hosts.1676329308196796999","timestamp":"2023-02-15T19:48:33.992Z"}
{"level":"debug","message":"Extracting Subgraph URL from artifact: xjoinindexpipeline.hosts.1676329308196796999","timestamp":"2023-02-15T19:48:33.998Z"}
{"level":"debug","message":"Extracted URL from artifact: http://xjoinindexpipeline-hosts-1676329308196796999.test.svc:4000/graphql","timestamp":"2023-02-15T19:48:33.998Z"}
{"level":"info","message":"Server ready at http://localhost:4001/graphql","timestamp":"2023-02-15T19:48:34.182Z"}
{"level":"http","message":"Supergraph query: query Hosts {\n  Hosts {\n    data {\n      display_name\n    }\n  }\n}","timestamp":"2023-02-15T19:48:40.859Z"}
```

### CODE REVIEW

File: src/logger/logger.ts [hunk 1]
```diff
@@ -9,13 +11,44 @@ const levels = {
...
+    return `${timestamp} ${level}: ${message}`;
+})
+
+export const logRequestMiddleware = (req: Request, res: Response, next: NextFunction) => {
+    if (req.method == 'OPTIONS') {
+        next()
+    } else if (req.body !== undefined && req.body.query !== undefined && req.body.query.indexOf('IntrospectionQuery') !== -1) {
+        next()
+    } else if (req !== undefined && req.body !== undefined) {
+        Logger.http(`Supergraph query: ${req.body.query}`)
```
Comment (thearifismail) [inline src/logger/logger.ts:38]: What does Logger.http() do? What format the message is logged int?

File: src/index.ts [hunk 2]
```diff
@@ -8,18 +8,18 @@ import { readFileSync } from "fs";
 import { Artifact, ArtifactsResponse } from "./interfaces";
 import { PREFIX } from "./registry";
 import express from 'express';
-import { Logger } from './logger/logger';
+import { Logger, logRequestMiddleware } from './logger/logger';
+import bodyParser from 'body-parser';
```
Comment (CodeHeeler) [inline src/index.ts:12]: Shouldn't curly braces be used around the import of bodyParser like for Logger above?

File: src/index.ts [hunk 3]
```diff
@@ -31,55 +31,54 @@ async function fetchArtifacts(): Promise<Response<ArtifactsResponse>> {
...
+
+    Logger.info("Fetching artifacts from Schema Registry")
     const artifactsRespose: Response<ArtifactsResponse> = await fetchArtifacts();
 
     const serviceDefinitions: ServiceDefinition[] = [];
     const artifacts: Artifact[] = artifactsRespose.body.artifacts;
 
     if (artifacts.length == 0) {
-        // TODO: INFO log here
+        Logger.warn("No artifacts found at the Schame Registry")
```
Comment (CodeHeeler) [inline src/index.ts:75]: Minor typo, I think you meant Schema Registry here.

Comment (strider) [reply src/index.ts:75]: What a shame! 😂

File: src/config/default.json [hunk 4]
```diff
@@ -5,5 +5,6 @@
         "Port": "1080"
     },
     "Port": "4000",
-    "LOG_LEVEL": "info"
+    "LOG_LEVEL": "info",
+    "GRAPHS_SYNC_TIMEOUT": "600000"
```
[the remark this answers is not in the document]
Comment (jpramos123) [reply src/config/default.json:9]: Yes, 10 minutes is too long!

At the end, this is a timeout that keeps running to fetch for new subgraphs periodically.  I don't see a reason to keep it running so often.

I will set it to 1 minute for now

File: src/index.ts [hunk 5]
```diff
@@ -8,18 +8,18 @@ import { readFileSync } from "fs";
...
 
     // Function to fetch artifacts from Schema Registry
     // In dev we use Apicurio
-
-    // TODO: DEBUG log here
-
-    return await got('v2/search/artifacts?labels=graphql', {
+    const artifactsPath = 'v2/search/artifacts?labels=graphql'
+    Logger.debug(`Fetching artifacts at: ${PREFIX}/${artifactsPath}`)
+    return await got(artifactsPath, {
```
[the remark this answers is not in the document]
Comment (jpramos123) [reply src/index.ts:22]: Yes, I will setup this!  [1]

Thanks

[1]https://github.com/sindresorhus/got/blob/main/documentation/6-timeout.md#timeout-options

File: src/index.ts [hunk 6]
```diff
@@ -95,11 +94,13 @@ async function loadSubgraphSchemas(): Promise<ServiceDefinition[]> {
 }
 
 const gateway = new ApolloGateway({
+
     async supergraphSdl({ update, healthCheck }) {
         const poll = function () {
-
+            Logger.info(`Fetching Subgraph Schemas every ${config.get("GRAPHS_SYNC_TIMEOUT")}ms`)
```
[the remark this answers is not in the document]
Comment (jpramos123) [reply src/index.ts:100]: Yes! Changing the constant name to `GRAPHS_SYNC_INTERVAL`

File: src/index.ts [hunk 7]
```diff
@@ -31,55 +31,54 @@ async function fetchArtifacts(): Promise<Response<ArtifactsResponse>> {
...
-        // TODO: DEBUG log here
         artifactPath = `v1/groups/${artifact.groupId}/artifacts/${artifact.id}`
+        Logger.debug(`Artifact has group, using resource path: ${artifactPath}`)
     } else {
-        // TODO: DEBUG log here
         artifactPath = `v1/artifacts/${artifact.id}`
+        Logger.debug(`Using resource path: ${artifactPath}`)
     }
 
+    Logger.debug(`Fetching artifact details at: ${PREFIX}/${artifactPath}`)
```
[the remark this answers is not in the document]
Comment (jpramos123) [reply src/index.ts:43]: In this case, we are just going to have 1 detail. But if you take a look at the `loadSubgraphSchemas()` method, we are logging the number of artifacts that were found.

Added error handlers and more logs to the code.

Thanks

File: src/logger/logger.ts (hunk 1 above)
Comment (jpramos123) [reply src/logger/logger.ts:38]: This is explicit telling that the application is receiving an HTTP request, in this case, we track requests that contains the SQL query in the body.

File: src/index.ts (hunk 2 above)
Comment (jpramos123) [reply src/index.ts:12]: No, in this specific case I want the object from the body-parser so I can use the following syntax `bodyParser.json()` to use the method  `json()` in from that object.

I would the curly braces import as `I'm getting stuff from **inside** this **place**`. 
I could do `import { json() }` in this specific case btw!

Also I'm removing this package from the import, we can use `express.json()` and it will produce the same effect

Thanks

File: src/index.ts [hunk 8]
```diff
@@ -65,39 +83,49 @@ function extractSubgraphUrl(artifact: Artifact) {
...
+                    typeDefs: parse(gqlResponse.body)
+                });
+            } catch (err : any) {
+                Logger.error(`Failed to fetch artifact details with error: ${err.message}`)
+                throw(err)
+            }
+        }
+    
+        return serviceDefinitions;
+    } catch (err : any) {
```
Comment (vkrizan) [inline src/index.ts:117]: This try-catch block is too large, and it has an inner catch within the for-loop. Mind to restructure it a bit?

File: src/registry.ts [hunk 9]
```diff
@@ -1,3 +1,5 @@
 import config from 'config';
 
 export const PREFIX = `${config.get('SchemaRegistry.Protocol')}://${config.get('SchemaRegistry.Hostname')}:${config.get('SchemaRegistry.Port')}/apis/registry`;
+
+export const REQUEST_TIMEOUT = {request: 3000}
```
Comment (vkrizan) [inline src/registry.ts:5]: No newline at the end of the file

File: src/config/default.json [hunk 10]
```diff
@@ -6,5 +6,5 @@
     },
     "Port": "4000",
     "LOG_LEVEL": "info",
-    "GRAPHS_SYNC_TIMEOUT": "600000" 
+    "GRAPHS_SYNC_INTERVAL": "60000"
```
Comment (vkrizan) [inline src/config/default.json:9]: Mind to explain this change? It seems unrelated.

**Review by vkrizan — COMMENTED** [review]
Thanks Joao, this would be very helpful.

Please ensure that JS statements are closed with a semicolon. I'd recommend to use some kind of a linter.

File: src/config/default.json (hunk 10 above)
Comment (jpramos123) [reply src/config/default.json:9]: 2 changes here.

First is the variable name, instead of using TIMEOUT use INTERVAL, as it is the interval the app is going to use to fetch data from the registry.

Second is the interval length, I was using 10 minutes, but looks like it could be too much, so I changed it to 1 minute

File: src/index.ts (hunk 8 above)
Comment (jpramos123) [reply src/index.ts:117]: Made some changes!

Thanks for the advise

File: src/index.ts [hunk 11]
```diff
@@ -1,127 +1,159 @@
...
+        Logger.debug(`Fetching artifacts at: ${PREFIX}/${artifactsPath}`);
+        const artifacts: Response<ArtifactsResponse> = await got.get(artifactsPath, {
+            prefixUrl: PREFIX,
+            responseType: "json",
+            headers: {
+                Accept: "application/json"
+            },
+            timeout: REQUEST_TIMEOUT
+        });
+        Logger.debug(`Successfuly fetched artifacts from: ${PREFIX}/${artifactsPath}`);
```
Comment (strider) [inline src/index.ts:32]: `s/Successfuly/Successfully/g`

File: src/index.ts [hunk 12]
```diff
@@ -1,127 +1,159 @@
...
-    });
+    try {
 
-    // TODO: DEBUG log here
+        Logger.debug(`Fetching artifact details at: ${PREFIX}/${artifactPath}`);
+        const artifact: Response<string> = await got.get(artifactPath, {
+            prefixUrl: PREFIX,
+            timeout: REQUEST_TIMEOUT,
+        });
+        Logger.debug(`Successfuly fetched artifact details from: ${PREFIX}/${artifactPath}`);
```
Comment (strider) [inline src/index.ts:59]: `s/Successfuly/Successfully/g`

**Review by strider — CHANGES_REQUESTED** [review]
Minor typo: `s/Successfuly/Successfully/g` otherwise looks good to me.

File: src/registry.ts [hunk 13]
```diff
@@ -1,3 +1,5 @@
-import config from 'config';
+import config from "config";
 
-export const PREFIX = `${config.get('SchemaRegistry.Protocol')}://${config.get('SchemaRegistry.Hostname')}:${config.get('SchemaRegistry.Port')}/apis/registry`;
+export const PREFIX = `${config.get("SchemaRegistry.Protocol")}://${config.get("SchemaRegistry.Hostname")}:${config.get("SchemaRegistry.Port")}/apis/registry`;
```
Comment (vkrizan) [inline src/registry.ts:3]: @jpramos123 are we sure that we want to keep this a constant? It uses config values, that seem like a moving part. Is config loaded before the constant is build? How is that ensured?

File: build_deploy_minikube.sh [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+git commit --amend
+#TAG=`cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 7 | head -n 1`
+TAG=$(git rev-parse --short=7 HEAD)
+#TAG="latest"
+IMAGE="127.0.0.1:5000/$APP"
+
+podman build -t $IMAGE:$TAG -f Dockerfile
+
+podman push $IMAGE:$TAG `minikube ip`:5000/$APP:$TAG --tls-verify=false
```
Comment (vkrizan) [inline build_deploy_minikube.sh:17]: There is lack of quotes for the variables and no error handling of `minikube ip`.

**Review by vkrizan — COMMENTED** [review]
Last two comments.

@jpramos123 this PR keeps growing with unrelated changes. I'd really suggest to keep the PR on topic and submit other changes separately to avoid this situation. That would help to have PRs merged quicker.

Thank you.

File: src/registry.ts (hunk 13 above)
Comment (jpramos123) [reply src/registry.ts:3]: I would prefer to keep this a constant. Also this is going to change when we are ready to use Apicurio in Clowder. Those values are going to be read from environment variables that are inside `clowdapp`.

Comment (vkrizan) [reply src/registry.ts]: Again, how we'd ensure here that the config is loaded before the constant is created? This is a red flag, as it could likely have an impact in future (e.g. expected this to change based on config, but it would use outdated values).

**Review by vkrizan — APPROVED** [review]
Approved. Feel free to improve the current PR or future one(s).

**Review by strider — APPROVED** [review]
Thanks! 👍
